document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresClient.java

```diff
@@ -0,0 +1,112 @@
+package org.hypertrace.core.documentstore.postgres;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class PostgresClient {
+
+  private static final Logger log = LoggerFactory.getLogger(PostgresClient.class);
+  private static final int VALIDATION_QUERY_TIMEOUT_SECONDS = 5;
+
+  private final String url;
+  private final String user;
+  private final String password;
+  private final int maxConnectionAttempts;
+  private final Duration connectionRetryBackoff;
+
+  private int count = 0;
+  private Connection connection;
+
+  public PostgresClient(
+      String url,
+      String user,
+      String password,
+      int maxConnectionAttempts,
+      Duration connectionRetryBackoff) {
+    this.url = url;
+    this.user = user;
+    this.password = password;
+    this.maxConnectionAttempts = maxConnectionAttempts;
+    this.connectionRetryBackoff = connectionRetryBackoff;
+  }
+
+  public synchronized Connection getConnection() {
+    try {
+      if (connection == null) {
+        newConnection();
+      } else if (!isConnectionValid(connection)) {
+        log.info("The database connection is invalid. Reconnecting...");
+        close();
+        newConnection();
+      }
+    } catch (SQLException sqle) {
+      throw new RuntimeException(sqle);
+    }
+    return connection;
+  }
+
+  private synchronized boolean isConnectionValid(Connection connection) {
```

Comment:
Do we need  `synchronized`  on `private` method? I think we should be fine without it as at any point in the time only one public entry to the getConnection function. what do you think?

Reply:
Just a symbolic one. One the public one is synchronized, it actually ensures all subsequent to private methods are synchronized as well. Just wanted to call it out.

Reply:
wouldn't it aquire its own lock? would it add to additional cost?